oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/InMemoryOAuth2AuthorizedClientService.java

```diff
@@ -80,7 +80,13 @@ public <T extends OAuth2AuthorizedClient> T loadAuthorizedClient(String clientRe
 		if (registration == null) {
 			return null;
 		}
-		return (T) this.authorizedClients.get(new OAuth2AuthorizedClientId(clientRegistrationId, principalName));
+		OAuth2AuthorizedClient cachedAuthorizedClient = this.authorizedClients
+				.get(new OAuth2AuthorizedClientId(clientRegistrationId, principalName));
+		if (cachedAuthorizedClient == null) {
+			return null;
+		}
+		return (T) new OAuth2AuthorizedClient(registration, cachedAuthorizedClient.getPrincipalName(),
+				cachedAuthorizedClient.getAccessToken(), cachedAuthorizedClient.getRefreshToken());
```

Comment:
Here we're no longer returning the instance that was saved in the map on `saveAuthorizedClient`, which means that if `T` is a subclass of `OAuth2AuthorizedClient`, this may cause a `ClassCastException`.
Is this an acceptable risk?
For context, the JDBC/R2DBC implementations can be configured with mappers to allow custom subtypes.

Reply:
I think the cast to `T` is still necessary given the contract of `OAuth2AuthorizedClientService`. It's not ideal, but the contract is what it is at this time. Do you see a way to avoid the cast?

Reply:
Unless we allow users to configure some sort of mapper to customize the type (like the JDBC/R2DBC implementations), I don't think we can avoid it.